allennlp/training/callbacks/callbacks.py

```diff
@@ -0,0 +1,512 @@
+                    self.tensorboard.add_train_scalar("current_batch_size", cur_batch)
+                    self.tensorboard.add_train_scalar("mean_batch_size", average)
+
+            if self.tensorboard.should_log_histograms_this_batch():
+                for name, param in trainer.model.named_parameters():
+                    self.param_updates[name].sub_(param.detach().cpu())
+                    update_norm = torch.norm(self.param_updates[name].view(-1, ))
+                    param_norm = torch.norm(param.view(-1, )).cpu()
+                    self.tensorboard.add_train_scalar("gradient_update/" + name,
+                                                      update_norm / (param_norm + 1e-7))
+                self.param_updates.clear()
+                self.tensorboard.log_histograms(trainer.model, self.histogram_parameters)
+
+        elif event == Events.EPOCH_END:
+            self.tensorboard.log_metrics(trainer.train_metrics,
+                                         val_metrics=trainer.val_metrics,
+                                         log_to_console=True,
+                                         epoch=trainer.epoch_number + 1)
+
+    @classmethod
+    def from_params(cls, serialization_dir: str, params: Params) -> 'LogToTensorboard':  # type: ignore
+        log_batch_size_period = params.pop_int("log_batch_size_period", None)
+        tensorboard = TensorboardWriter.from_params(params=params,
+                                                    serialization_dir=serialization_dir,
+                                                    get_batch_num_total=lambda: None)
+        return LogToTensorboard(tensorboard, log_batch_size_period)
+
+
+@Callback.register("learning_rate_scheduler")
+class LrsCallback(Callback['trainer2.Trainer']):
+    """
+    Callback that handles the learning rate scheduler.
+    """
+    def __init__(self, learning_rate_scheduler: LearningRateScheduler) -> None:
+        self.learning_rate_scheduler = learning_rate_scheduler
+
+    def __call__(self, event: str, trainer: 'trainer2.Trainer') -> None:
+        if event == Events.AFTER_BACKWARD:
+            self.learning_rate_scheduler.step_batch(trainer.batch_num_total)
+        elif event == Events.EPOCH_END:
+            self.learning_rate_scheduler.step(trainer.latest_val_metric, trainer.epoch_number)
+
+    def get_training_state(self) -> dict:
+        """
+        We need to persist the learning_rate_scheduler state as training state.
+        """
+        return {"learning_rate_scheduler": self.learning_rate_scheduler.state_dict()}
+
+    def restore_training_state(self, training_trainer: dict) -> None:
+        state_dict = training_trainer.pop("learning_rate_scheduler", None)
+
+        if state_dict:
+            self.learning_rate_scheduler.load_state_dict(state_dict)
+
+
+
+    @classmethod
+    def from_params(cls, params: Params, optimizer: Optimizer) -> 'LrsCallback':  # type: ignore
```

Comment:
You really shouldn't need this one - doesn't the optimizer just get passed through as an extra?